man/bapbundle.1
.TH "BAPBUNDLE" 1 "2015-01-26" "BAP" "Bap Manual"
.SH NAME
bapbundle \- manages BAP bundles.
.SH SYNOPSIS
.P
.nf
\fBbapbundle\fR \-help
\fBbapbundle\fR \fICOMMAND\fR \-help
\fBbapbundle\fR \fBinstall\fR [\fIOPTION\fR]...  \fIBUNDLE\fR
\fBbapbundle\fR \fBremove\fR  [\fIOPTION\fR]...  \fIBUNDLE\fR
\fBbapbundle\fR \fBshow\fR    [\fIOPTION\fR]...  \fIBUNDLE\fR
\fBbapbundle\fR \fBupdate\fR  [\fIOPTION\fR]...  \fIBUNDLE\fR
\fBbapbundle\fR \fBpack\fR    [\fIOPTION\fR]...  \fIBUNDLE\fR
.fi

.SH DESCRIPTION
.P

A bundle is an opaque package containing anything that is requested for a
program to work in a standalone mode. The \fBbapbundle\fR tool allows a user
to create bundles, update meta information and pack resources. It also can
install and remove bundles into the designated bundle repository, where they
can be found by a main BAP program without extra options.

Usually the \fBbapbundle\fR command is used implicitly by the
\fBbapbuild\fR, so there is no need to use the \fBpack\fR command
directly. A user can, however, add some resource files or update meta
information, using the \fBupdate\fR command.

The \fBinstall\fR command copies a \fIBUNDLE\fR to a known
destination, that is seearched by a main program, when plugins are
loaded automatically. The default destination is specified at program
configuration time and usually works out of box. The destination can
be overridden in the runtime, however.

The \fBremove\fR command removes a previously installed \fIBUNDLE\fR
from a given destination.

.SH ARGUMENTS
.TP 4
\fIBUNDLE\fR
Name of a bundle.

.SH COMMON OPTIONS
.TP 4
\fB\-help\fR, \fB\-\-help\fR
Prints a brief summary of command line options.


.SH COMMON META OPTIONS
.P

The following options sets or updates meta information stored in the
\fIBUNDLE\fR.

.TP 4
\fB\-author\fR \fINAME\fR

Set author's name. If unspecified, then a value of \fBUSER\fR
environment variable is used. Ideally, the name should be of the
form \fBFull Name [Short name] <email>\fR

.TP 4
\fB\-name\fR \fINAME\fR

Sets the name of the \fIBUNDLE\fR to the \fINAME\fR. If not specified,
then a basename without an extension of the \fINAME\fR will be used.

.TP 4
\fB\-desc\fR \fITEXT\f

Sets the \fITEXT\fR as a package description. The text should consist
of a one line short description, followed by an optional detailed
multiline description. The detailed description should be separated
from the short one by a new line symbol.

.SH INSTALL AND REMOVE OPTIONS

.TP
\fB\-destdir\fR \fIDIR\f

Specifies a directory for \fBinstall\fR and \fBremove\fR commands.
If unspecified, then a default directory specified at configuration
time will be used.

.SH SHOW OPTIONS

.TP 4
\fB\-m\FR

Shows meta information.

.TP 4
\fB\-r\FR

Prints a list of resource files stored in the \fIBUNDLE\fR.

.TP 4
\fB\-l\FR

Prints a list of libraries and other loadable code, stored in the
\fIBUNDLE\fR.

.SH UPDATE OPTIONS
.P
Other then \fBCOMMON META OPTIONS\fR the \fBupdate\fR command accepts
the following options:

.TP 4
\fB\-add-resources \fIRESOURCE\-LIST\fR

Adds resources specified by the \fIRESOURCE\-LIST\fR to the
\fIBUNDLE\fR.  The \fIRESOURCE\-LIST\fR is a comma separated list of
resources, where each resources is either a file path or a named
filepath of a form \fB<name>=<path>\fR, where \fB<name>\fR defines
how a resource will accessed in the bundle. If not specified, then
the \fB<name>\fR is equal to \fB<path>\fR. The \fIRESOURCE\-LIST\fR
mustn't contain spaces, unless a filename contains it. In that case
the resource list must be enclosed in parenthesis.

The \fIRESOURCE\-LIST\fR grammar:
.P
.nf
.RS
resources ::= resource [, resources]
resource  ::= path | name=path
.RE
.fi

.SH PACK OPTIONS
.P
Other then \fBCOMMON META OPTIONS\fR the \fBupdate\fR command accepts
the following options:


.TP 4
\fB\-main\fR \fINAME\fR

Set the entry point to \fINAME\fR. The entry point is a module that is
loaded by a bap main program.

.TP 4
\fB\-requires\fR \fIFEATURE-LIST\fR

Define package requirements. The requirements are libraries used by
the package. The \fIFEATURE-LIST\fR must be a comma separated list, of
requirements. Each requirement is a library name, optionally followed
by filename of the implementation. If implementation is provided, then
it will be included in the \fIBUNDLE\fR. Othrewise, only the name will
be stored, and the implementation will be searched at the time of
loading using \fBocaml-findlib\fR.

The \fIFEATURE\-LIST\fR grammar:
.P
.RS
.nf
feature-list ::= feature [, feature]
feature      ::= name | name=path
.fi
.RE

.TP 4
\fB\-provides\fR \fIFEATURE-LIST\fR

Specify a list of features provided by the package. The
\fIFEATURE-LIST\fR should a comma separated list of names, where each
name is an interface name provided by a plugin.

.TP 4
\fB\-resources\fR \fRESOURCES\fR

Specify a list of resources used by the package. See
\fB\-add-resource\fR option for more information.

.SH EXIT STATUS
.P

Program returns:
   \- 0 if everything is ok.
   \- 1 for malformed command line
   \- 2 for all other errors.

.SH ENVIRONMENT
.P

Program reads \fBUSER\fR variable to get a user name.

.SH BUGS
.P
Report bugs to https://github.com/BinaryAnalysisPlatform/bap/issues

.SH SEE ALSO
.P
\fBbap\fR(1),\fBbapbuild\fR(1),\fBocamlfind\fR(1)


CHANGES.md
1.0.0-rc1
=========

1. A more powerful plugin system

   A plugin now is an archive, like a Java's jar file, or Mac OS X
   bundle. The plugin can contain libraries, that are provided by the
   plugin, libraries on which a plugin depends (optionally), and
   arbitrary resources. Plugins, are compiled with `bapbuild`, bundled
   and installed with `bapbundle` tool. The installed plugins are
   loaded automatically. Most of the functionality is now provided via
   plugin system and many injection points are added to the library.
   For example, llvm loader and disassembler, all lifters, byteweight
   are all now just plugins. Also, the new plugin system is much more
   safe, as it keeps track of loaded interfaces and prevents state
   corruption.

   Also, plugins now work in a toplevel.


2. Split Bap.Std into several libraries.
   The following new libraries were added or carved out of the Bap.Std,
   or bap frontend.

   - Regular - supporting code for regular data types.
     Implements IO (new Data framework), Regular and Opaque interfaces,
     and a new Caching interface.
   - Graphlib - is a separete library that doesn't depend on BAP
   - Dwarf - parse DWARF infromation
   - Elf - parse ELF files
   - Byteweight - find function starts
   - Traces - new library for loading execution traces
   - ARM - ARM lifting
   - X86_cpu - X86 support library with CPU declarations
   - Demangle - a small library for demangling names
   - Future - a new library to work with co-inductive values
   - Ida - a library to interact with IDA
   - Piqi - a serialization library based on piqi
   - Text-tags - a library for adding semantic tags to formatters

3. The disassembler layer is severly rewritten

   Now we represent a CFG as a Graplib's graph, instead of `block
   table`.  The `Symtab` module is now a collection of CFG, so the
   interface is quite different. The `Disasm` module now is also quite
   different, as we cleaned up the inteface, and removed functions that
   weren't very popular but influenced on performace and memory footprint.

   The disassembly level is simplified, there is only one CFG now, with only
   one instruction type. The instruction type now stores all semantic properties in
   a bitset. A user can add its own properties to the instruction.

   New reconstruction algorithm is now more precise, but slightly
   slower.  The reason lies mostly in the LLVM backend, that was
   imprecise, in his semantic information. Now we don't trust LLVM blindly,
   but instead use our lifters to provide a better information about instruction
   semantics.

4. Made project storable and loadable

   The project data structure now can be saved and loaded. The whole
   state will be correctly restored, including a state stored on
   module level.  A new caching framework is used to cache the
   project, that gives a significant speed up, on a usual use case -
   runing passes over a project.

5. Added new injection points

   BAP now heavily relies on a Dependecy injection principle. The
   library itself tries not to provide implementations, but instead
   specifies interfaces. The implementation can be provided by a user
   using the plugin system. A set of reasonable default
   implementations, is of course provided. Currently there the following
   injection points provided by Bap.Std and other libraries in the platform:

   - project pass - arbitrary program transformations or computations;
   - serialization - provide your own serialization formats;
   - printing - change or add new printing formats;
   - loader - support new input formats
   - disassembler - implement your own disassembler;
   - targets - to provide lifters
   - ABI - implement different ABI
   - symbolizer - name functions
   - rooter - find function starts
   - reconstructor - CFG reconstruction algorithms
   - caching - provide custom support for caching data

   We provide the following set of plugins by default:

   - emit-ida-script  extract a python script from the project data type
   - dump-symbols     dump symbol information as a list of blocks
   - arm              provide ARM lifter
   - piqi-printers    provides piqi serialization for main data types (BIL, IR)
   - phoenix          output project information in a phoenix format
   - print            print project in various formats
   - cache            provide caching services
   - byteweight       find function starts using Byteweight algorithm
   - x86              provide x86 lifter
   - llvm             provide loader and disassembler
   - ida              use ida to provide rooter, symbolizer and reconstructor
   - read-symbols     read symbol information from file
   - elf-loader       read ELF and DWARF formats in a pure OCaml


6. Added BIL interpreters

   We formally specified BIL semantics and implemented it with several
   extensible interpreters:
   - expi - interpreter for an expression sublanguage of BIL
   - bili - BIL interpreter
   - biri - IR interpreter

7. Removed bap-server



0.9.9
=====

1. Graphlib library

   `Graphlib` is a generic library that extends a well known
   OCamlGraph library. `Graphlib` uses its own, more reach and modern,
   `Graph` interface that is isomorphic to OCamlGraph's `Sigs.P`
   signature for persistant graphs. This interface is developed
   according to the Janestreet's style guidlines and depends on
   Core_kernel library.  Other than the new interface, `Graphlib`
   provides several graph implementations, and generic algorithms.  To
   make our algorithms polymorphic over chosen graph representation we
   use first-class modules, instead of functors, that makes library
   syntactically more light weight.

2. Refined IR.

   phi and arg terms were refined. a phi term now is a discriminated
   set of expressions, and arguments are made more like a defintions.

3. SSA form

   A transformation to a semipruned SSA form was added to a library

4. Changed Var interface.


tools/bapbundle.ml
open Core_kernel.Std
open Bap_bundle.Std
open Format

open Manifest.Fields

exception Provides_field_syntax_error
exception Target_unspecified
exception Target_is_directory
exception Target_doesn't_exist

let target = ref ""
let manifest = ref (Manifest.create "")
let modified = String.Hash_set.create ()
let destdir = ref Bap_config.libdir

let destdir_arg =
  "-destdir", Arg.Set_string destdir,
  sprintf "DIR Set the destination directory (default: %s)"
    !destdir

let field_modified field =
  Hash_set.add modified (Field.name field)

let is_field_modified field =
  Hash_set.mem modified (Field.name field)

let set field value =
  field_modified field;
  manifest := Field.fset field !manifest value

let add field x =
  let xs = Field.get field !manifest in
  match List.mem xs x with
  | true -> ()
  | false ->
    field_modified field;
    manifest := Field.fset field !manifest (xs @ [x])

let set_target t = target := t

let chop_extension name =
  if String.mem name '.'
  then Filename.chop_extension name
  else name

let name_of_path path =
  Filename.basename path |> chop_extension

let fields = String.split ~on:','

let named_resource r = match String.split ~on:'=' r with
  | [name;file] -> Some name, file
  | _ -> None,r

let open_bundle () = match target.contents with
  | "" -> raise Target_unspecified
  | x when Sys.is_directory x -> raise Target_is_directory
  | x when Sys.file_exists x -> Bundle.of_uri (Uri.of_string x)
  | x -> raise Target_doesn't_exist


module Update = struct
  let resources = ref []

  let add_res r =
    resources := r :: !resources

  let common_args = Arg.[
      "-author",    String (set author), "<name> Set bundle's author name";
      "-name",      String (set name),   "<name> Set bundle's name";
      "-desc",      String (set desc),   "<text> Set bundle's description";
    ]

  let args = common_args @ Arg.[
      "-add-resources", String add_res,"<list> Add resource files";
    ]

  let update m field =
    if is_field_modified field
    then Field.fset field m (Field.get field !manifest)
    else m

  let main () =
    let bundle = open_bundle () in
    Bundle.update_manifest bundle ~f:(fun init ->
        List.fold [author;name;desc] ~init ~f:update);
    List.map !resources ~f:named_resource |>
    List.map ~f:(fun (name,path) -> name, Uri.of_string path) |>
    Bundle.insert_files bundle
end

module Pack = struct
  let pack = Bundle.Builder.create ()

  let named_library f = match String.split ~on:'=' f with
    | [name;file] -> Some (Some name,file)
    | _ -> None


  let bundle_file (name,uri) =
    Bundle.Builder.put_file pack ?name (Uri.of_string uri)

  let bundle_files files = List.iter files ~f:bundle_file

  let add_requires fs =
    List.iter (fields fs) ~f:(fun f -> match String.split ~on:'=' f with
        | [name] -> add requires name
        | [name;path] ->
          add requires name;
          bundle_file (None,path)
        | _ -> raise Provides_field_syntax_error)

  let add_provides fs = List.iter (fields fs) ~f:(add provides)

  let add_resources rs =
    List.map (fields rs) ~f:named_resource |>
    bundle_files

  let set_main file =
    set main (Filename.chop_extension file);
    bundle_file (None,file)

  let args = Update.common_args @ Arg.[
      "-main",      String (set_main),      "<name> Set the entry point";
      "-requires",  String (add_requires),  "<list> List of required artifacts";
      "-provides",  String (add_provides),  "<list> List of provided interfaces";
      "-resources", String (add_resources), "<list> List of embedded resources";
    ]

  let main () =
    let manifest = !manifest in
    let target = if target.contents = ""
      then Manifest.name manifest
      else target.contents in
    if target = "" then raise Target_unspecified;
    let uri = Uri.of_string target in
    Bundle.Builder.embed_manifest pack manifest;
    Bundle.Builder.flush pack uri
end

module Show = struct
  let smanifest = ref false
  let resources = ref false
  let libraries = ref false

  let args = Arg.[
      "-m", Set smanifest, "Show meta information";
      "-r", Set resources, "Show available resources";
      "-l", Set libraries, "Show packed libraries"
    ]

  let show_manifest m =
    printf "%s@." @@ Manifest.to_string m

  let show_files k m b =
    let choose = match k with
      | `libraries when libraries.contents -> ident
      | `resources when resources.contents -> not
      | _ -> fun _ -> false in
    let is_library name =
      Filename.check_suffix name ".cmxs" ||
      Filename.check_suffix name ".cma" in
    List.iter (Bundle.list b) ~f:(fun name ->
        if choose (is_library name)
        then printf "%s@." name)

  let main () =
    let bundle = open_bundle () in
    let m = Bundle.manifest bundle in
    if smanifest.contents then show_manifest m;
    show_files `resources m bundle;
    show_files `libraries m bundle
end

module Install = struct
  let args = [destdir_arg]

  let main () =
    if target.contents = "" then raise Target_unspecified;
    if Sys.file_exists !target
    then if Sys.is_directory !target
      then raise Target_is_directory
      else FileUtil.cp [!target] !destdir
    else raise Target_doesn't_exist
end

module Remove = struct
  let args = [destdir_arg]
  let main () =
    if target.contents = "" then raise Target_unspecified;
    let file = Filename.concat !destdir !target in
    if Sys.file_exists file
    then Sys.remove file
end


let args = ref []

let parse = ref ignore
let main = ref ident

module type Command = sig
  val args : (string * Arg.spec * string) list
  val main : unit -> unit
end

let select arg =
  let switch (module Command : Command) =
    args := Arg.align (Command.args @ !args);
    parse := set_target;
    main := Command.main in
  match arg with
  | "pack"    -> switch (module Pack)
  | "update"  -> switch (module Update)
  | "show"    -> switch (module Show)
  | "install" -> switch (module Install)
  | "remove"  -> switch (module Remove)
  | msg -> raise (Arg.Bad ("Unknown command: " ^ msg))

let dispatch arg = !parse arg

let usage_msg = "USAGE

  bapbundle -help
  bapbundle <command> -help
  bapbundle install [<install-options>] <bundle>
  bapbundle remove  [<remove-options>]  <bundle>
  bapbundle show    [<show-options>]    <bundle>
  bapbundle update  [<update-options>]  <bundle>
  bapbundle pack    [<pack-options>]    <bundle>

DESCRIPTION

  Manages BAP bundles. If you're looking for a way to
  create a bundle, then bapbuild is the tool that you're
  looking for.

  See bapbundle(1) man-page for a full manual.

OPTIONS
"

let error fmt =
  kfprintf (fun ppf -> pp_print_newline ppf (); exit 2) err_formatter fmt

let usage fmt =
  kfprintf (fun ppf ->
      pp_print_newline ppf ();
      pp_print_newline ppf ();
      pp_print_string ppf @@
      Arg.usage_string !args usage_msg;
      pp_print_newline ppf ();
      exit 1)
    err_formatter fmt


let () =
  parse := select;
  Arg.parse_dynamic args dispatch usage_msg;
  try !main () with
  | Provides_field_syntax_error ->
    error "An entry in provides list should be <name> | <name>=<path>"
  | Target_unspecified ->
    usage "Please specify <bundle>"
  | Target_is_directory ->
    error "The <bundle> is a directory"
  | Target_doesn't_exist ->
    error "Can't open <bundle> file, it doesn't exist"
  | Not_a_bundle ->
    error "The provided <bundle> file is either malformed or not a bundle"
  | Sys_error err ->
    error "%s" err
  | exn ->
    error "Unexpected error: %s\nPlease report" @@
    Exn.to_string exn
